aws-lc-rs/src/rsa/encryption.rs

```diff
@@ -0,0 +1,510 @@
+        let key = generate_rsa_key(size.bits(), false)?;
+        Self::new(key)
+    }
+
+    /// Generate a RSA `KeyPair` of the specified key-strength.
+    ///
+    /// Supports the following key sizes:
+    /// * `EncryptionKeySize::Rsa2048`
+    /// * `EncryptionKeySize::Rsa3072`
+    /// * `EncryptionKeySize::Rsa4096`
+    ///
+    /// # Errors
+    /// * `Unspecified`: Any key generation failure.
+    #[cfg(feature = "fips")]
+    pub fn generate_fips(size: KeySize) -> Result<Self, Unspecified> {
+        let key = generate_rsa_key(size.bits(), true)?;
+        Self::new(key)
+    }
+
+    /// Construct a `PrivateDecryptingKey` from the provided PKCS#8 (v1) document.
+    ///
+    /// Supports RSA key sizes between 2048 and 8192 (inclusive).
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs during deserialization of this key from PKCS#8.
+    pub fn from_pkcs8(pkcs8: &[u8]) -> Result<Self, KeyRejected> {
+        let key = encoding::pkcs8::decode_der(pkcs8)?;
+        Ok(Self::new(key)?)
+    }
+
+    /// Returns a boolean indicator if this RSA key is an approved FIPS 140-3 key.
+    #[cfg(feature = "fips")]
+    #[must_use]
+    pub fn is_valid_fips_key(&self) -> bool {
+        super::key::is_valid_fips_key(&self.0)
+    }
+
+    /// Returns the RSA key size in bytes.
+    #[must_use]
+    pub fn key_size_bytes(&self) -> usize {
+        key_size_bytes(&self.0)
+    }
+
+    /// Returns the RSA key size in bits.
+    #[must_use]
+    pub fn key_size_bits(&self) -> usize {
+        key_size_bits(&self.0)
+    }
+
+    /// Retrieves the `PublicEncryptingKey` corresponding with this `PrivateDecryptingKey`.
+    #[must_use]
+    #[allow(clippy::missing_panics_doc)]
+    pub fn public_key(&self) -> PublicEncryptingKey {
+        PublicEncryptingKey::new(self.0.clone()).expect(
+            "PublicEncryptingKey key size to be supported by PrivateDecryptingKey key sizes",
+        )
+    }
+}
+
+impl Debug for PrivateDecryptingKey {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_tuple("PrivateDecryptingKey").finish()
+    }
+}
+
+impl AsDer<Pkcs8V1Der<'static>> for PrivateDecryptingKey {
+    fn as_der(&self) -> Result<Pkcs8V1Der<'static>, Unspecified> {
+        Ok(Pkcs8V1Der::new(encoding::pkcs8::encode_v1_der(&self.0)?))
+    }
+}
+
+impl Clone for PrivateDecryptingKey {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+/// An RSA public key used for encrypting plaintext that is decrypted by a [`PrivateDecryptingKey`].
+pub struct PublicEncryptingKey(LcPtr<EVP_PKEY>);
+
+impl PublicEncryptingKey {
+    fn new(evp_pkey: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
+        Self::validate_key(&evp_pkey)?;
+        Ok(Self(evp_pkey))
+    }
+
+    fn validate_key(key: &LcPtr<EVP_PKEY>) -> Result<(), Unspecified> {
+        if !is_rsa_key(key) {
+            return Err(Unspecified);
+        };
+        match key_size_bits(key) {
```

Comment:
Is size the only thing worth checking? What about calling AWS-LC's validate which does a few more checks for the public/private key.

Reply:
AWS-LC already perform the key validation checks when we parse a private or public key, per @dkostic we didn't need to perform any additional checks.